### ISSUE TITLE
Massive test failures when multiple PR builders are started in parallel

### ISSUE THREAD

**Issue by Donnerbart** [hazelcast/hazelcast#11991]
Multiple test failures happen, when several PR builders are started in parallel (e.g. when rebasing multiple branches and pushing them within a short period of time).

It would be good to check if there are any unwanted interactions between the builds (network, machine-thread limit, CPU load, some other shared resource). Maybe we should have some monitoring of critical resources like CPU, memory usage, process count, thread count on the build server?

```
com.hazelcast.spi.impl.operationservice.impl.Invocation_DetectHeartbeatTimeoutTest.whenExpiresEventually
com.hazelcast.map.MapLockTest.testLockMigration
com.hazelcast.map.impl.mapstore.MapStoreWriteThroughTest.testOneMemberWriteThroughWithLRU
com.hazelcast.map.impl.mapstore.MapStoreWriteBehindTest.testIssue1085WriteBehindBackupWithLongRunnigMapStore
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsAll_whenMapCreatedInEager
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoads_whenMapLazyAndCheckingSize
com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testDoesntLoadAgain_whenLoaderNodeGoesDown
com.hazelcast.map.BackupTest.testGracefulShutdown
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMultipleSlaveMembers_withPartitionInitialization
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMemberAndCluster_concurrently_withPartitionInitialization
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMasterAndTerminateSlaveMember_concurrently
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownAndTerminateSlaveMembers_concurrently
com.hazelcast.internal.cluster.impl.MembershipUpdateTest.memberListOrder_shouldBeSame_whenMemberRestartedWithSameIdentity
com.hazelcast.executor.SingleNodeTest.invokeAllTimeoutSuccess
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/12321/testReport/

```
com.hazelcast.concurrent.lock.ConditionAdvancedTest.testShutDownNode_whenOtherWaitingOnConditionAwait
com.hazelcast.concurrent.lock.LockAdvancedTest.testLockOwnerDies
com.hazelcast.concurrent.lock.LockAdvancedTest.testKeyOwnerDies
com.hazelcast.internal.cluster.impl.MembershipUpdateTest.memberListOrder_shouldBeSame_whenMemberRestartedWithSameIdentity
com.hazelcast.internal.cluster.impl.MembershipUpdateTest.com.hazelcast.internal.cluster.impl.MembershipUpdateTest
com.hazelcast.internal.cluster.impl.NoMigrationClusterStateTest.rebalancing_shouldNotHappen_whenMemberLeaves
com.hazelcast.internal.cluster.impl.NoMigrationClusterStateTest.lostPartitions_shouldBeAssigned_toNewMembers
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownAndTerminateSlaveMembers_concurrently
com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.partitionAssignmentsShouldBeCorrect_whenClusterIsSafe
com.hazelcast.map.AsyncTest.testSetAsyncWithTtl
com.hazelcast.map.MapLockTest.testLockTTLKey
com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoading
com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedWhileLoadingAndNoBackups
com.hazelcast.map.impl.mapstore.MapLoaderFailoverTest.testLoadsAll_whenInitialLoaderNodeRemovedAfterLoading
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testDoesNotLoadAgain_whenLoadedLazyAndNodeAdded
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsNothing_whenMapCreatedLazy
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsAll_whenMapCreatedInEager
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsMap_whenLazyAndValueRetrieved
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsAll_whenLazyModeAndLoadAll
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadsOnce_whenSizeCheckedTwice
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoads_whenMapLazyAndCheckingSize
com.hazelcast.map.impl.mapstore.MapStoreWriteThroughTest.testOneMemberWriteThroughWithLRU
com.hazelcast.map.impl.nearcache.MapNearCacheSerializationCountTest.testSerializationCounts[method:GET mapFormat:OBJECT nearCacheFormat:BINARY invalidateOnChange:true serializeKeys:false]
com.hazelcast.map.impl.tx.TxnMapNearCacheSerializationCountTest.testSerializationCounts[method:GET mapFormat:BINARY nearCacheFormat:BINARY serializeKeys:false]
com.hazelcast.map.impl.tx.TxnMapNearCacheSerializationCountTest.testSerializationCounts[method:GET mapFormat:BINARY nearCacheFormat:OBJECT serializeKeys:false]
com.hazelcast.map.impl.tx.TxnMapNearCacheSerializationCountTest.testSerializationCounts[method:GET mapFormat:OBJECT nearCacheFormat:OBJECT serializeKeys:false]
com.hazelcast.internal.cluster.impl.NoMigrationClusterStateTest.lostPartitions_shouldBeAssigned_toAvailableMembers
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testDoesNotLoadAgain_whenLoadedAndNodeAdded
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMemberAndCluster_concurrently_withPartitionInitialization
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMultipleSlaveMembers_withPartitionInitialization
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMasterAndTerminateSlaveMember_concurrently
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/12322/testReport/

```
com.hazelcast.internal.cluster.impl.NoMigrationClusterStateTest.lostPartitions_shouldBeAssigned_toAvailableMembers
com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.clusterShouldBeEventuallySafe_duringMigration
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testLoadAgain_whenLoadedAllCalledMultipleTimes
com.hazelcast.map.impl.mapstore.MapLoaderMultiNodeTest.testDoesNotLoadAgain_whenLoadedAndNodeAdded
com.hazelcast.map.impl.mapstore.MapStoreTest.testGetAllKeys
com.hazelcast.spi.impl.operationservice.impl.Invocation_TimeoutTest.whenMultipleThreadsCallGetOnSameLongRunningOperation
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMasterAndTerminateSlaveMember_concurrently
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMemberAndCluster_concurrently_withPartitionInitialization
com.hazelcast.internal.partition.GracefulShutdownTest.shutdownMultipleSlaveMembers_withPartitionInitialization
com.hazelcast.map.BackupTest.testGracefulShutdown
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/12323/testReport/

```
com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.clusterShouldBeEventuallySafe_duringMigration
com.hazelcast.spi.impl.operationservice.impl.Invocation_OnMemberLeftTest.whenMemberLeaves
com.hazelcast.spi.impl.operationservice.impl.Invocation_OnMemberLeftTest.whenMemberRestarts_withSameAddress
com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl_invokeOnPartitionsTest.testLongRunning
com.hazelcast.test.mocknetwork.MockJoinerTest.parallelJoin
com.hazelcast.internal.cluster.impl.MembershipUpdateTest.memberListOrder_shouldBeSame_whenMemberRestartedWithSameIdentity
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/12324/testReport/

**Comment by Donnerbart** [comment]
I saved the build logs from S3, before they are buried in the glacier:
[new-lab-fast-pr-12321.zip](https://github.com/hazelcast/hazelcast/files/1559209/new-lab-fast-pr-12321.zip)
[new-lab-fast-pr-12322.zip](https://github.com/hazelcast/hazelcast/files/1559210/new-lab-fast-pr-12322.zip)
[new-lab-fast-pr-12323.zip](https://github.com/hazelcast/hazelcast/files/1559211/new-lab-fast-pr-12323.zip)
[new-lab-fast-pr-12324.zip](https://github.com/hazelcast/hazelcast/files/1559212/new-lab-fast-pr-12324.zip)

**Comment by Donnerbart** [comment]
All builds seem to have quite a lot of GC pauses, e.g.:
```
11:51:30, accumulated pauses: 2472 ms, max pause: 120 ms, pauses over 1000 ms: 0
11:51:35, accumulated pauses: 3718 ms, max pause: 448 ms, pauses over 1000 ms: 0
11:51:40, accumulated pauses: 3117 ms, max pause: 470 ms, pauses over 1000 ms: 0
11:51:45, accumulated pauses: 2628 ms, max pause: 173 ms, pauses over 1000 ms: 0
11:51:50, accumulated pauses: 2633 ms, max pause: 625 ms, pauses over 1000 ms: 0
11:51:55, accumulated pauses: 2152 ms, max pause: 441 ms, pauses over 1000 ms: 0
11:52:00, accumulated pauses: 2190 ms, max pause: 446 ms, pauses over 1000 ms: 0
11:52:05, accumulated pauses: 8207 ms, max pause: 5667 ms, pauses over 1000 ms: 2
11:52:10, accumulated pauses: 1280 ms, max pause: 841 ms, pauses over 1000 ms: 0
11:52:15, accumulated pauses: 3477 ms, max pause: 549 ms, pauses over 1000 ms: 0
11:52:20, accumulated pauses: 4023 ms, max pause: 2567 ms, pauses over 1000 ms: 1
11:52:25, accumulated pauses: 1846 ms, max pause: 141 ms, pauses over 1000 ms: 0
11:52:30, accumulated pauses: 3464 ms, max pause: 301 ms, pauses over 1000 ms: 0
11:52:35, accumulated pauses: 2399 ms, max pause: 550 ms, pauses over 1000 ms: 0
11:52:40, accumulated pauses: 2502 ms, max pause: 251 ms, pauses over 1000 ms: 0
11:52:45, accumulated pauses: 3202 ms, max pause: 877 ms, pauses over 1000 ms: 0
11:52:50, accumulated pauses: 1697 ms, max pause: 397 ms, pauses over 1000 ms: 0
11:52:55, accumulated pauses: 1477 ms, max pause: 602 ms, pauses over 1000 ms: 0
11:53:00, accumulated pauses: 1491 ms, max pause: 176 ms, pauses over 1000 ms: 0
11:53:05, accumulated pauses: 3136 ms, max pause: 178 ms, pauses over 1000 ms: 0
11:53:10, accumulated pauses: 1871 ms, max pause: 416 ms, pauses over 1000 ms: 0
11:53:15, accumulated pauses: 2467 ms, max pause: 820 ms, pauses over 1000 ms: 0
11:53:20, accumulated pauses: 2401 ms, max pause: 191 ms, pauses over 1000 ms: 0
11:53:25, accumulated pauses: 2650 ms, max pause: 572 ms, pauses over 1000 ms: 0
11:53:30, accumulated pauses: 3268 ms, max pause: 266 ms, pauses over 1000 ms: 0
11:53:35, accumulated pauses: 2371 ms, max pause: 149 ms, pauses over 1000 ms: 0
11:53:40, accumulated pauses: 1872 ms, max pause: 109 ms, pauses over 1000 ms: 0
11:53:45, accumulated pauses: 1891 ms, max pause: 161 ms, pauses over 1000 ms: 0
11:53:50, accumulated pauses: 3014 ms, max pause: 300 ms, pauses over 1000 ms: 0
11:53:55, accumulated pauses: 2052 ms, max pause: 118 ms, pauses over 1000 ms: 0
11:54:00, accumulated pauses: 1680 ms, max pause: 97 ms, pauses over 1000 ms: 0
11:54:05, accumulated pauses: 3172 ms, max pause: 423 ms, pauses over 1000 ms: 0
11:54:10, accumulated pauses: 2416 ms, max pause: 145 ms, pauses over 1000 ms: 0
11:54:15, accumulated pauses: 1750 ms, max pause: 103 ms, pauses over 1000 ms: 0
11:54:20, accumulated pauses: 2735 ms, max pause: 182 ms, pauses over 1000 ms: 0
11:54:25, accumulated pauses: 4827 ms, max pause: 3188 ms, pauses over 1000 ms: 1
11:54:30, accumulated pauses: 1050 ms, max pause: 307 ms, pauses over 1000 ms: 0
```

Looks like we had 1829 out of 8336 pauses over 1 second, that's almost 22%.
```
pauses over 1000 ms: 0 (78.06 %)
6507
pauses over 1000 ms: 1 (18.93 %)
1578
pauses over 1000 ms: 2 ( 2.92 %)
245
pauses over 1000 ms: 3 ( 0.05 %)
4
pauses over 1000 ms: 4 ( 0.02 %)
2
```

**Comment by Donnerbart** [comment]
The `GracefulShutdownTest` seems to be failing quite often under these circumstances.

**Comment by Donnerbart** [comment]
I ran in the same situation again and had closer look at what the `GracefulShutdownTest` does. I creates a bunch of Hazelcast instances in each of its 30 test methods. This is a bit uncommonly written, but should not be so much different from creating a small cluster in the `@Before` method (in terms of the total count of Hazelcast instances).

But there was a spinning method without any wait time, which could hog a thread quite extensively. I'm preparing a PR to add a wait time here. I'm also thinking of removing the `ParallelTest`, since this test seems to be a bit sensitive when executed with too many tasks in parallel.

**Comment by kwart** [comment]
Should be addressed by #12832, so I'm closing this issue.
Feel free to reopen if needed.
